### PR TITLE
Document stateless components in formal types

### LINKED ISSUE #6624: Formal Type Definitions not up-to-date
Hey,

I noticed "Formal Type Definitions" are not up-to-date as they don't account for stateless components.

It probably should be something like:

```
type ReactComponentElement<TProps> = {
  type : ReactClass<TProps> | ReactStatelessComponent<TProps>,
  props : TProps,
  key : string | boolean | number | null,
  ref : string | null
};
```

https://github.com/facebook/react/blob/master/docs/docs/ref-10-glossary.md

### PR DESCRIPTION
**Pull request by sheerun**
Fixes #6624

### CODE REVIEW

**Comment by facebook-github-bot-2** [comment]
Thank you for your pull request and welcome to our community. We require contributors to sign our Contributor License Agreement, and we don't seem to have you on file. In order for us to review and merge your code, please sign up at https://code.facebook.com/cla - and if you have received this in error or have any questions, please drop us a line at cla@fb.com. Thanks!

**Comment by facebook-github-bot-8** [comment]
Thank you for signing our Contributor License Agreement. We can now accept your code for this (and any) Facebook open source project. Thanks!

File: docs/docs/ref-10-glossary.md [hunk 1]
```diff
@@ -137,6 +137,7 @@ This is why you shouldn't construct your own instance. Instead, `ReactElement` i
 
 The `render` method of a `ReactComponent` is expected to return another `ReactElement`. This allows these components to be composed. Ultimately the render resolves into `ReactElement` with a `string` tag which instantiates a DOM `Element` instance and inserts it into the document.
 
+React 0.14 introduced [stateless functional components](https://facebook.github.io/react/blog/2015/10/07/react-v0.14.html#stateless-functional-components) as an alternative way of defining components. Indeaf of being a class, it is a simple function that accepts props and is expected to return a `ReactElement`.
```
Comment (gaearon) [inline docs/docs/ref-10-glossary.md]: Typo, should be `Instead`

File: docs/docs/ref-10-glossary.md [hunk 2]
```diff
@@ -189,5 +190,7 @@ type ReactComponent<TProps> = {
   props : TProps,
   render : () => ReactElement
 };
+
+type ReactStatelessComponent<TProps> = (TProps) => ReactElement;
```
Comment (gaearon) [inline docs/docs/ref-10-glossary.md]: I wonder if we want to call this `ReactFunctionalStatelessComponent` or `ReactFunctionalComponent` instead. It’s stateless alright, but class components can also be stateless. Not sure.

Comment (sheerun) [inline docs/docs/ref-10-glossary.md]: I'd leave it as it's the same as in the source:
https://github.com/facebook/react/blob/master/src/renderers/shared/reconciler/__tests__/ReactStatelessComponent-test.js

Comment (gaearon) [inline docs/docs/ref-10-glossary.md]: The internal naming can sometimes differ from what we want to expose publicly (e.g. we call shared modules `isomorphic` in the source but we never use this word in the docs). So maybe this is one of those cases.

cc @sebmarkbage for terminology clarification

**Comment by sheerun** [comment]
Fixed the typo

**Comment by facebook-github-bot-2** [comment]
@sheerun updated the pull request.

**Comment by gaearon** [comment]
Another thing we don’t currently document here is that you can return `null` from `render()` and, since 15, from functional components. This is probably an omission.

Maybe something like

```
type ReactNodeList = ReactNode | ReactEmpty | undefined;

type ReactEmpty = null | boolean;

---

type ReactComponent<TProps> = {
  props : TProps,
  render : () => ReactElement | ReactEmpty
};

type ReactStatelessComponent<TProps> = (TProps) => ReactElement | ReactEmpty;
```

would be more correct.

**Comment by sheerun** [comment]
So returning undefined is deprecated, and stateless component can return boolean, right?

**Comment by facebook-github-bot-3** [comment]
@sheerun updated the pull request.

**Comment by gaearon** [comment]
I don't think returning undefined was ever supported. It is in ReactEmpty because ReactEmpty was not used as return value. Essentially returning null or false doesn't seem documented today at all. I'm just proposing one possible way to fix this.

File: docs/docs/ref-10-glossary.md (hunk 2 above)
Comment (sebmarkbage) [inline docs/docs/ref-10-glossary.md]: It should be ReactFunctionalComponent. Our internal naming is wrong.

**Comment by gaearon** [comment]
I’d say let’s not tackle `null` here, it’s a separate issue.
Please change `ReactStatelessComponent` to `ReactFunctionalComponent`, and this is good to go.

Thanks. 👍

**Comment by facebook-github-bot-1** [comment]
@sheerun updated the pull request.

File: docs/docs/ref-10-glossary.md [hunk 3]
```diff
@@ -137,6 +137,7 @@ This is why you shouldn't construct your own instance. Instead, `ReactElement` i
 
 The `render` method of a `ReactComponent` is expected to return another `ReactElement`. This allows these components to be composed. Ultimately the render resolves into `ReactElement` with a `string` tag which instantiates a DOM `Element` instance and inserts it into the document.
 
+React 0.14 introduced [stateless functional components](https://facebook.github.io/react/blog/2015/10/07/react-v0.14.html#stateless-functional-components) as an alternative way of defining components. Instead of being a class, it is a simple function that accepts props and is expected to return a `ReactElement`.
```
Comment (gaearon) [inline docs/docs/ref-10-glossary.md]: Missed this: we need the link to start with `/react/...` in case we ever change the domain and for local dev.

**Comment by facebook-github-bot-4** [comment]
@sheerun updated the pull request.

**Comment by gaearon** [comment]
Thank you!

**Pull request merged**